app/src/org/commcare/utils/AppInstallationUtil.java

```diff
@@ -0,0 +1,123 @@
+package org.commcare.utils;
+
+import android.content.Context;
+import android.widget.Toast;
+
+import org.commcare.CommCareApp;
+import org.commcare.CommCareApplication;
+import org.commcare.activities.CommCareSetupActivity;
+import org.commcare.engine.resource.AppInstallStatus;
+import org.commcare.google.services.analytics.AnalyticsParamValue;
+import org.commcare.google.services.analytics.FirebaseAnalyticsUtil;
+import org.commcare.resources.model.Resource;
+import org.commcare.tasks.ConnectionDiagnosticTask;
+import org.commcare.tasks.ResourceEngineTask;
+import org.commcare.tasks.templates.CommCareTaskConnector;
+import org.commcare.views.notifications.MessageTag;
+import org.commcare.views.notifications.NotificationMessageFactory;
+import org.javarosa.core.services.locale.Localization;
+
+import static org.commcare.CommCareNoficationManager.AIRPLANE_MODE_CATEGORY;
+import static org.commcare.activities.CommCareSetupActivity.DIALOG_INSTALL_PROGRESS;
+import static org.commcare.activities.CommCareSetupActivity.INSTALL_MODE_BARCODE;
+import static org.commcare.activities.CommCareSetupActivity.INSTALL_MODE_FROM_LIST;
+import static org.commcare.activities.CommCareSetupActivity.INSTALL_MODE_MANAGED_CONFIGURATION;
+import static org.commcare.activities.CommCareSetupActivity.INSTALL_MODE_OFFLINE;
+import static org.commcare.activities.CommCareSetupActivity.INSTALL_MODE_SMS;
+import static org.commcare.activities.CommCareSetupActivity.INSTALL_MODE_URL;
+
+/**
+ * @author $|-|!˅@M
+ */
+public class AppInstallationUtil {
+
+    public static void startResourceInstall(Context context, String reference,
+                                            CommCareTaskConnector connector, CommCareApp ccApp,
+                                            boolean shouldSleep, int lastInstallMode) {
+        if (lastInstallMode == INSTALL_MODE_OFFLINE) {
+            startResourceInstall(reference, connector, ccApp, shouldSleep, lastInstallMode);
+        } else {
+            ConnectionDiagnosticTask<CommCareSetupActivity> task = new ConnectionDiagnosticTask(context);
```

Comment:
would be good to explicity typecast `new ConnectionDiagnosticTask(context)` as `new ConnectionDiagnosticTask<>(context)`